### ISSUE TITLE
Container HEALTHCHECKs can lead to hanging API calls and bad State

### ISSUE THREAD

**Issue by jahkeup** [moby/moby#36661]
**Description**

Occasionally the Docker daemon stops responding to interactions with a containers that have HEALTHCHECKs. The problem presents itself in [several older versions of Docker and the latest packaged versions on Ubuntu and v17.12.0-ce on Amazon Linux](https://github.com/jahkeup/health-stats-repro#tested-against).

By our observations, this looks to be a race condition that is met with a deadlock that prevents further calls against affected containers.

This observed issue may be related to https://github.com/moby/moby/issues/35933 . I'm working on bisecting the releases (using https://github.com/docker/docker-ce) using this repro to narrow down the problem further in any case.

**Steps to reproduce the issue:**

A repro case has been [built and run against several version of docker](https://github.com/jahkeup/health-stats-repro) with positive results after a few rounds of execution (I recommend 10-20 rounds to tickle the bug). There likely isn't anything specific about the 2 containers, but it has been positively triggering the bug for this test.

1. Build container image with HEALTHCHECK defined (`echo hello` every `1s`)
2. Start 2 containers using image
3. Wait some time (`10s` in our test)
4. Stop containers
5. Inspect containers

**Describe the results you received:**

Started containers appear to continue running and to be healthy despite being non-responsive. 

```bash
ubuntu@ip-172-31-37-156:~$ docker ps
CONTAINER ID        IMAGE                      COMMAND               CREATED             STATUS                    PORTS               NAMES
0cf518c205f7        docker-poke:healthchecks   "sh -c 'sleep 30m'"   17 minutes ago      Up 17 minutes (healthy)                       sad_hugle
ubuntu@ip-172-31-37-156:~$ docker inspect 0cf518c205f7
^C
```

Additionally, the output of `docker ps`  will continue reporting that the container is still up and running even though the process will exit after 30m (started with `sleep 30m`).

```bash
0cf518c205f7        docker-poke:healthchecks   "sh -c 'sleep 30m'"   37 minutes ago      Up 37 minutes (healthy)                               sad_hugle
```

**Describe the results you expected:**

I expected that I would be able to inspect this container.

```bash
docker inspect 0cf518c205f7
{
   ...
}
```

**Additional information you deem important (e.g. issue happens only occasionally):**

This issue is readily made apparent with a few concurrent runs, but otherwise lies dormant even with many serial runs.

**Output of `docker version`:**

```
Client:
 Version:       17.12.1-ce
 API version:   1.35
 Go version:    go1.9.4
 Git commit:    7390fc6
 Built: Tue Feb 27 22:17:40 2018
 OS/Arch:       linux/amd64

Server:
 Engine:
  Version:      17.12.1-ce
  API version:  1.35 (minimum version 1.12)
  Go version:   go1.9.4
  Git commit:   7390fc6
  Built:        Tue Feb 27 22:16:13 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 5
Server Version: 17.12.1-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 4
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9b55aab90508bd389d7654c4baf173a981477d55
runc version: 9f9c96235cc97674e935002fc3d78361b696a69e
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-1052-aws
Operating System: Ubuntu 16.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.625GiB
Name: ip-172-31-37-156
ID: L4W4:V4WA:OHSS:QTGL:DRJG:32GX:7DKK:FFLO:WKR2:IJYV:NKDG:GWRA
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Ubuntu 16.04 on AWS EC2 using the [repro runner](https://github.com/jahkeup/health-stats-repro#ubuntu)

Thanks @samuelkarp!

| Package                   | Result |
|---------------------------|--------|
| `17.09.1~ce-0~ubuntu`     | pass   |
| `17.10.0~ce-0~ubuntu`     | pass   |
| `17.11.0~ce-0~ubuntu`     | pass   |
| `17.12.0~ce~rc1-0~ubuntu` | fail   |
| `17.12.0~ce-0~ubuntu`     | fail   |
| `17.12.1~ce-0~ubuntu`     | fail   |
| `18.01.0~ce-0~ubuntu`     | fail   |
| `18.02.0~ce-0~ubuntu`     | fail   |
| `18.03.0~ce~rc4-0~ubuntu` | fail   |

Amazon Linux on AWS EC2 using the [repro runner](https://github.com/jahkeup/health-stats-repro#amazon-linux)

Thanks @jhaynes!

| Package      | Result |
|--------------|--------|
| `17.12.0-ce` | fail   |
| `17.09.1-ce` | pass   |

**Comment by stevvooe** [comment]
@jahkeup Great reproduction!

As far as I can tell, it sounds like a classic dead lock. To save us some time, would it be possible for you to grab a stack trace from the when the issue is encountered on your reproduction? If I can't solve it with that, I'll dig into the reproduction and work from there.

To grab the stack trace, run the daemon with debug mode enabled, trigger the bug, then send the `SIGUSR1` to the daemon. The instructions for doing this are at https://docs.docker.com/config/daemon/#enable-debugging.

**Comment by jahkeup** [comment]
I'll get my docker into debug mode and dump this again, however I was able to collect a dump from the instance I used for the above issue content as-is.

[non-debug-stuck: goroutine-stacks-2018-03-21T213731Z.log](https://github.com/moby/moby/files/1835347/goroutine-stacks-2018-03-21T213731Z.log)

**Comment by stevvooe** [comment]
@jahkeup Thanks for being responsive!

I took a peak at the stack trace and it looks like the lock is being held by `Kill`. Here is the relevant stack trace:

```
goroutine 3709 [select, 1 minutes]:
github.com/docker/docker/vendor/google.golang.org/grpc/transport.(*Stream).Header(0xc4219baa20, 0x2aee0e8, 0xc421db6b58, 0x38b4a60)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/transport/transport.go:256 +0x146
github.com/docker/docker/vendor/google.golang.org/grpc.recvResponse(0x7f59f32150b0, 0xc420e72d20, 0xc4203a00c0, 0xc4203a00e0, 0x38ab220, 0x392bc30, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/call.go:64 +0xac
github.com/docker/docker/vendor/google.golang.org/grpc.invoke(0x7f59f32150b0, 0xc420e72d20, 0x2a9618d, 0x28, 0x28fd420, 0xc420e72c60, 0x295f380, 0x392bc30, 0xc4200709c0, 0x0, ...)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/call.go:279 +0xcfe
github.com/docker/docker/vendor/github.com/containerd/containerd.namespaceInterceptor.unary(0x2a4fa1c, 0x4, 0x7f59f320d030, 0xc420018028, 0x2a9618d, 0x28, 0x28fd420, 0xc420e72c60, 0x295f380, 0x392bc30, ...)
	/go/src/github.com/docker/docker/vendor/github.com/containerd/containerd/grpc.go:18 +0xf4
github.com/docker/docker/vendor/github.com/containerd/containerd.(namespaceInterceptor).(github.com/docker/docker/vendor/github.com/containerd/containerd.unary)-fm(0x7f59f320d030, 0xc420018028, 0x2a9618d, 0x28, 0x28fd420, 0xc420e72c60, 0x295f380, 0x392bc30, 0xc4200709c0, 0x2aee0c0, ...)
	/go/src/github.com/docker/docker/vendor/github.com/containerd/containerd/grpc.go:34 +0xf4
github.com/docker/docker/vendor/google.golang.org/grpc.Invoke(0x7f59f320d030, 0xc420018028, 0x2a9618d, 0x28, 0x28fd420, 0xc420e72c60, 0x295f380, 0x392bc30, 0xc4200709c0, 0x0, ...)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/call.go:141 +0xdd
github.com/docker/docker/vendor/github.com/containerd/containerd/api/services/tasks/v1.(*tasksClient).Kill(0xc420aabf58, 0x7f59f320d030, 0xc420018028, 0xc420e72c60, 0x0, 0x0, 0x0, 0xc421add770, 0xc421422e60, 0x10)
	/go/src/github.com/docker/docker/vendor/github.com/containerd/containerd/api/services/tasks/v1/tasks.pb.go:468 +0xd2
github.com/docker/docker/vendor/github.com/containerd/containerd.(*task).Kill(0xc421add770, 0x38aba20, 0xc420018028, 0x9, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/containerd/containerd/task.go:180 +0x206
github.com/docker/docker/libcontainerd.(*client).SignalProcess(0xc420128a80, 0x38aba20, 0xc420018028, 0xc421284100, 0x40, 0x2a4f90c, 0x4, 0x9, 0xc421edae40, 0xc4218ccd80)
	/go/src/github.com/docker/docker/libcontainerd/client_daemon.go:375 +0xd6
github.com/docker/docker/daemon.(*Daemon).kill(0xc4203ca200, 0xc421b518c0, 0x9, 0x2, 0x2)
	/go/src/github.com/docker/docker/daemon/kill.go:179 +0x8c
github.com/docker/docker/daemon.(*Daemon).killWithSignal(0xc4203ca200, 0xc421b518c0, 0x9, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/kill.go:99 +0x243
github.com/docker/docker/daemon.(*Daemon).killPossiblyDeadProcess(0xc4203ca200, 0xc421b518c0, 0x9, 0x144ca49, 0x0)
	/go/src/github.com/docker/docker/daemon/kill.go:169 +0x4c
github.com/docker/docker/daemon.(*Daemon).Kill(0xc4203ca200, 0xc421b518c0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/kill.go:129 +0x82
github.com/docker/docker/daemon.(*Daemon).ContainerKill(0xc4203ca200, 0xc4218ccc11, 0x40, 0x0, 0x392d8a0, 0x8400000002912fe0)
	/go/src/github.com/docker/docker/daemon/kill.go:52 +0x83
github.com/docker/docker/api/server/router/container.(*containerRouter).postContainersKill(0xc42087e280, 0x7f59f32150b0, 0xc420e72ba0, 0x38a9d20, 0xc421274540, 0xc421c39a00, 0xc420e72ae0, 0x2a508c0, 0x5)
	/go/src/github.com/docker/docker/api/server/router/container/container_routes.go:210 +0xed
github.com/docker/docker/api/server/router/container.(*containerRouter).(github.com/docker/docker/api/server/router/container.postContainersKill)-fm(0x7f59f32150b0, 0xc420e72ba0, 0x38a9d20, 0xc421274540, 0xc421c39a00, 0xc420e72ae0, 0x7f59f32150b0, 0xc420e72ba0)
	/go/src/github.com/docker/docker/api/server/router/container/container.go:62 +0x69
github.com/docker/docker/api/server/middleware.ExperimentalMiddleware.WrapHandler.func1(0x7f59f32150b0, 0xc420e72ba0, 0x38a9d20, 0xc421274540, 0xc421c39a00, 0xc420e72ae0, 0x7f59f32150b0, 0xc420e72ba0)
	/go/src/github.com/docker/docker/api/server/middleware/experimental.go:27 +0xd8
github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1(0x7f59f32150b0, 0xc420e72b40, 0x38a9d20, 0xc421274540, 0xc421c39a00, 0xc420e72ae0, 0xc421db98a8, 0x1186c47)
	/go/src/github.com/docker/docker/api/server/middleware/version.go:62 +0x423
github.com/docker/docker/pkg/authorization.(*Middleware).WrapHandler.func1(0x7f59f32150b0, 0xc420e72b40, 0x38a9d20, 0xc421274540, 0xc421c39a00, 0xc420e72ae0, 0x10, 0x1176655)
	/go/src/github.com/docker/docker/pkg/authorization/middleware.go:59 +0x83b
github.com/docker/docker/api/server/middleware.DebugRequestMiddleware.func1(0x7f59f32150b0, 0xc420e72b40, 0x38a9d20, 0xc421274540, 0xc421c39a00, 0xc420e72ae0, 0x7f59f32150b0, 0xc420e72b40)
	/go/src/github.com/docker/docker/api/server/middleware/debug.go:25 +0x967
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1(0x38a9d20, 0xc421274540, 0xc421c39a00)
	/go/src/github.com/docker/docker/api/server/server.go:137 +0x1c6
net/http.HandlerFunc.ServeHTTP(0xc42089f2c0, 0x38a9d20, 0xc421274540, 0xc421c39a00)
	/usr/local/go/src/net/http/server.go:1918 +0x44
github.com/docker/docker/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc420012d20, 0x38a9d20, 0xc421274540, 0xc421c39a00)
	/go/src/github.com/docker/docker/vendor/github.com/gorilla/mux/mux.go:103 +0x22b
github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP(0xc420a49df0, 0x38a9d20, 0xc421274540, 0xc421c39a00)
	/go/src/github.com/docker/docker/api/server/router_swapper.go:29 +0x70
net/http.serverHandler.ServeHTTP(0xc42006de10, 0x38a9d20, 0xc421274540, 0xc421c39a00)
	/usr/local/go/src/net/http/server.go:2619 +0xb4
net/http.(*conn).serve(0xc420a85540, 0x38ab9e0, 0xc42150d140)
	/usr/local/go/src/net/http/server.go:1801 +0x71d
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2720 +0x288
```

I think the next step here would be to start investigating what is going on in containerd. We need to figure out whether the request is making it there or not.

**Comment by jahkeup** [comment]
Was able to reproduce this with the newly released  `18.03.0-ce` docker daemon on Ubuntu 16.04.

I've collected info, version, and the stack traces for this daemon for a stuck shim:

[18.03.0-ce-ubuntu1604-36661-20180322.tar.gz](https://github.com/moby/moby/files/1838751/18.03.0-ce-36661-20180322.tar.gz)

**Comment by stevvooe** [comment]
Looking at the shim goroutine dump, we can see that we are waiting on a fifo to start the process:

```
goroutine 91 [select, 9 minutes]:
github.com/containerd/containerd/vendor/github.com/containerd/fifo.OpenFifo(0x7fb85deaf0a0, 0xc420050180, 0xc420166fd0, 0xa3, 0x1, 0xc400000000, 0x300000002, 0xc420125200, 0xc42012f780, 0x432c4f)
	/go/src/github.com/containerd/containerd/vendor/github.com/containerd/fifo/fifo.go:122 +0x381
github.com/containerd/containerd/linux/proc.copyPipes(0x807d80, 0xc420050180, 0x8083a0, 0xc4200713a0, 0x0, 0x0, 0xc420166fd0, 0xa3, 0xc420167080, 0xa3, ...)
	/go/src/github.com/containerd/containerd/linux/proc/io.go:53 +0x3be
github.com/containerd/containerd/linux/proc.(*execProcess).start(0xc4201683c0, 0x807d80, 0xc420050180, 0x0, 0x0)
	/go/src/github.com/containerd/containerd/linux/proc/exec.go:167 +0x591
github.com/containerd/containerd/linux/proc.(*execCreatedState).Start(0xc42008e308, 0x807d80, 0xc420050180, 0x0, 0x0)
	/go/src/github.com/containerd/containerd/linux/proc/exec_state.go:40 +0xa2
github.com/containerd/containerd/linux/proc.(*execProcess).Start(0xc4201683c0, 0x807d80, 0xc420050180, 0x40, 0xc420096bd8)
	<autogenerated>:1 +0x51
github.com/containerd/containerd/linux/shim.(*Service).Start(0xc42000e0c0, 0x807d80, 0xc420050180, 0xc420076c20, 0x0, 0x0, 0x0)
```

Looking at all other goroutines locked in `runtime_SemacquireMutex` on 0xc42000e0c4, this is the only one that is holding a lock and not trying to acquire the semaphore.

**Comment by jahkeup** [comment]
The attached output from the above shim has a deleted FIFO for its stdout:

```
root@ip-172-31-47-105:/proc/1286/fd# lsof -p 1286
COMMAND    PID USER   FD      TYPE             DEVICE SIZE/OFF    NODE NAME
docker-co 1286 root  cwd       DIR               0,19      140    1849 /run/docker/containerd/daemon/io.containerd.runtime.v1.linux/moby/18d45667967f6c29dd86ae36b85130416f1afefa6537bbdbc66a5f5a029c58ad
docker-co 1286 root  rtd       DIR              259,1     4096       2 /
docker-co 1286 root  txt       REG              259,1  4324256   51295 /usr/bin/docker-containerd-shim
docker-co 1286 root    0r      CHR                1,3      0t0       6 /dev/null
docker-co 1286 root    1u     unix 0xffff8803a851ec00      0t0 1463760 type=STREAM
docker-co 1286 root    2u     unix 0xffff8803a851ec00      0t0 1463760 type=STREAM
docker-co 1286 root    4u  a_inode               0,11        0    8025 [eventpoll]
docker-co 1286 root    5u  a_inode               0,11        0    8025 [eventpoll]
docker-co 1286 root    6u     unix 0xffff880360544800      0t0 2019410 @/containerd-shim/moby/18d45667967f6c29dd86ae36b85130416f1afefa6537bbdbc66a5f5a029c58ad/shim.sock type=STREAM
docker-co 1286 root    7u     unix 0xffff880387a35c00      0t0 2016007 @/containerd-shim/moby/18d45667967f6c29dd86ae36b85130416f1afefa6537bbdbc66a5f5a029c58ad/shim.sock type=STREAM
docker-co 1286 root    8r     FIFO               0,10      0t0 2016009 pipe
docker-co 1286 root    9r     FIFO               0,10      0t0 2088440 pipe
docker-co 1286 root   10r     FIFO               0,10      0t0 2016010 pipe
docker-co 1286 root   12r     FIFO               0,10      0t0 2016011 pipe
docker-co 1286 root   20w     FIFO               0,10      0t0 2088440 pipe
docker-co 1286 root   21r     FIFO               0,10      0t0 2088441 pipe
docker-co 1286 root   22u     FIFO               0,19      0t0     640 /run/docker/containerd/18d45667967f6c29dd86ae36b85130416f1afefa6537bbdbc66a5f5a029c58ad/be86934cf866422f105bb87b7b4938d5445b23d1f44aa8a6dee909793393a865-stdout (deleted)
docker-co 1286 root   23r     FIFO               0,10      0t0 2088442 pipe
```

There's no difference between the above where the container is attempted to be stopped and removed, and the below where the container is started and then inspected (no stop or remove attempts):

```
COMMAND    PID USER   FD      TYPE             DEVICE SIZE/OFF   NODE NAME
docker-co 1331 root  cwd       DIR               0,19      140   1515 /run/docker/containerd/daemon/io.containerd.runtime.v1.linux/moby/42930b0c577475cb4a77e8be0d54f0e8a8eb026c4ccff14d69b4e68b80e89619
docker-co 1331 root  rtd       DIR              259,1     4096      2 /
docker-co 1331 root  txt       REG              259,1  4324256  51295 /usr/bin/docker-containerd-shim
docker-co 1331 root    0r      CHR                1,3      0t0      6 /dev/null
docker-co 1331 root    1w      CHR                1,3      0t0      6 /dev/null
docker-co 1331 root    2w      CHR                1,3      0t0      6 /dev/null
docker-co 1331 root    3r     FIFO               0,10      0t0 390245 pipe
docker-co 1331 root    4u  a_inode               0,11        0   8025 [eventpoll]
docker-co 1331 root    5u  a_inode               0,11        0   8025 [eventpoll]
docker-co 1331 root    6u     unix 0xffff880391a29c00      0t0  80731 @/containerd-shim/moby/42930b0c577475cb4a77e8be0d54f0e8a8eb026c4ccff14d69b4e68b80e89619/shim.sock type=STREAM
docker-co 1331 root    7u     unix 0xffff880394c01c00      0t0  80735 @/containerd-shim/moby/42930b0c577475cb4a77e8be0d54f0e8a8eb026c4ccff14d69b4e68b80e89619/shim.sock type=STREAM
docker-co 1331 root    8r     FIFO               0,10      0t0  80737 pipe
docker-co 1331 root    9w     FIFO               0,10      0t0 390245 pipe
docker-co 1331 root   10r     FIFO               0,10      0t0  80738 pipe
docker-co 1331 root   12r     FIFO               0,10      0t0  80739 pipe
docker-co 1331 root   20r     FIFO               0,10      0t0 390246 pipe
docker-co 1331 root   21u     FIFO               0,19      0t0   3262 /run/docker/containerd/42930b0c577475cb4a77e8be0d54f0e8a8eb026c4ccff14d69b4e68b80e89619/9b200a3c4f15d44fd029699fcca2d5ccece43f2c7dc6d5edeb198e4eb4ec251b-stdout (deleted)
docker-co 1331 root   22r     FIFO               0,10      0t0 390247 pipe
```

**Comment by stevvooe** [comment]
https://github.com/containerd/containerd/pull/2229 is a possible fix for this, or it at least alleviates the problem.

I still think something is awry with the healthchecks. There may be a rarer condition that is exasperated when the FIFOs block.

**Comment by stevvooe** [comment]
We are kicking off a release of containerd 1.0.3 in https://github.com/containerd/containerd/pull/2242 with the mitigation. I think there is a more complete fix on the docker side to prevent this condition from happening.

**Comment by stevvooe** [comment]
Note that I've also added https://github.com/moby/moby/pull/36751 to mitigate problems with healthchecks. There may be a few more fixes that follow.

**Comment by jahkeup** [comment]
I've run the changes in #36751 (without applying the containerd mitigation) using the same repro as before and am seeing the "limiter wait" log message when the concurrent checks would have run. The issue seems to be resolved with this change and the expected behavior is observed. Are there any other lock holders or interactions with a container's fifos that could trigger the same behavior (eg: an `exec` during a `HEALTHCHECK`)?

**Comment by stevvooe** [comment]
@jahkeup Thanks for checking that!

> Are there any other lock holders or interactions with a container's fifos that could trigger the same behavior (eg: an exec during a HEALTHCHECK)?

Probably, although #36751 will avoid the cascade failure that happens by overwhelming the kernel with fifo/exec creation that can happen when running a lot of health checks.

**Comment by jahkeup** [comment]
@stevvooe the latest changes in https://github.com/moby/moby/pull/36751/commits/cda213b065e051090ec7908a379f874fe93715be are passing the repro test :+1:

**Comment by spex66** [comment]
as it seems that I've run into this one, may I've the question:
=> Is there a way to start/stop the containers without risking to lose data?
=> more general, how to get control of a docker deployment which is in this state?
thx

**Comment by cpuguy83** [comment]
@spex66 

> Is there a way to start/stop the containers without risking to lose data?

No data is lost from stopping/starting containers other than in-memory state

> more general, how to get control of a docker deployment which is in this state?

Fix a borken daemon? Restart dockerd.

More permanent fix: Upgrade to 18.03.1.

---
 
@jahkeup Is this resolved on your end with 18.03.1?

**Comment by jahkeup** [comment]
@cpuguy83 The issue appears to be resolved with 18.03.1 (tested on Amazon Linux and Ubuntu) :+1:
